src/vm/analysis/update_expressions_sorting/mod.rs

```diff
@@ -0,0 +1,304 @@
+                                    Some(list) => &list[0],
+                                    _ => &function_args[0]
+                                };
+                                if let Some(tle_name) = define_expr.match_atom() {
+                                    return Some((tle_name.clone(), define_expr.id, define_expr));
+                                }   
+                            }
+                        }
+                        _ => {}
+                    }
+                } 
+            } 
+        }
+        None
+    }
+}
+
+pub struct TopLevelExpressionIndex {
+    expr_index: usize,
+    atom_index: u64
+}
+
+struct Graph {
+    adjacency_list: Vec<Vec<usize>>
+}
+
+impl Graph {
+    fn new() -> Self {
+        Self { adjacency_list: Vec::new() }
+    }
+
+    fn add_node(&mut self, expr_index: usize) {
+        self.adjacency_list.push(vec![]);
+    }
+
+    fn add_directed_edge(&mut self, src_expr_index: usize, dst_expr_index: usize) {
+        let list = self.adjacency_list.get_mut(src_expr_index).unwrap();
+        list.push(dst_expr_index);
+    }
+    
+    fn get_node_descendants(&self, expr_index: usize) -> Vec<usize> {
+        self.adjacency_list[expr_index].clone()
+    }
+
+    fn has_node_descendants(&self, expr_index: usize) -> bool {
+        self.adjacency_list[expr_index].len() > 0
+    }
+
+    fn nodes_count(&self) -> usize { self.adjacency_list.len() }
+}
+
+struct GraphWalker {
+    seen: HashSet<usize>,
+}
+
+impl GraphWalker {
+
+    fn new() -> Self { Self { seen: HashSet::new() } }
+
+    /// Depth-first search producing a post-order sort
+    fn get_sorted_dependencies(&mut self, graph: &Graph) -> CheckResult<Vec<usize>> {
+        let mut sorted_indexes = Vec::<usize>::new();
+        for expr_index in 0..graph.nodes_count() {
+            self.sort_dependencies_recursion(expr_index, graph, &mut sorted_indexes);
+        }
+
+        Ok(sorted_indexes)
+    }
+
+    fn sort_dependencies_recursion(&mut self, tle_index: usize, graph: &Graph, branch: &mut Vec<usize>) {
+        if self.seen.contains(&tle_index) {
+            return
+        }
+
+        self.seen.insert(tle_index);
+        if let Some(list) = graph.adjacency_list.get(tle_index) {
+            for neighbor in list.iter() {
+                self.sort_dependencies_recursion(neighbor.clone(), graph, branch);
+            }
+        }
+        branch.push(tle_index);
+    }
+
+    fn get_cycling_dependencies(&mut self, graph: &Graph, sorted_indexes: &Vec<usize>) -> Option<Vec<usize>> {
+        let mut tainted: HashSet<usize> = HashSet::new();
+
+        for node in sorted_indexes.iter() {
+            let mut tainted_descendants_count = 0;
+            let descendants = graph.get_node_descendants(*node);
+            for descendant in descendants.iter() {
+                if graph.has_node_descendants(*descendant) == false || tainted.contains(descendant) {
```

Comment:
`!graph.has_node_descendants(*descendant)`

Reply:
Addressed in https://github.com/blockstack/blockstack-core/pull/1079/commits/fd23f90f92281830fb2b4d2708ea01a041e66cf0